proxy/src/connection.rs

```diff
@@ -133,32 +133,31 @@ impl BoundPort {
                 .and_then(move |socket| {
                     let remote_addr = socket.peer_addr()
                         .expect("couldn't get remote addr!");
+
                     // TODO: On Linux and most other platforms it would be better
                     // to set the `TCP_NODELAY` option on the bound socket and
                     // then have the listening sockets inherit it. However, that
                     // doesn't work on all platforms and also the underlying
                     // libraries don't have the necessary API for that, so just
                     // do it here.
                     set_nodelay_or_warn(&socket);
-                    let why_no_tls = if let Some((_identity, config_watch)) = &tls {
-                        // TODO: use `identity` to differentiate between TLS
-                        // that the proxy should terminate vs. TLS that should
-                        // be passed through.
-                        if let Some(config) = &*config_watch.borrow() {
-                            let f = tls::Connection::accept(socket, config.clone())
-                                .map(move |tls| {
-                                    (Connection::tls(tls), remote_addr)
-                                });
-                            return Either::A(f);
-                        } else {
-                            // No valid TLS configuration.
-                            tls::ReasonForNoTls::NoConfig
-                        }
-                    } else {
-                        tls::ReasonForNoTls::Disabled
+
+                    let conn = match tls::current_connection_config(&tls) {
+                        Conditional::Some(config) => {
+                            // TODO: use `config.identity` to differentiate
+                            // between TLS that the proxy should terminate vs.
+                            // TLS that should be passed through.
+                            let f = tls::Connection::accept(socket, config.config)
+                                .map(move |tls| Connection::tls(tls));
+                            Either::A(f)
+                        },
+                        Conditional::None(why_no_tls) => {
+                            let f = future::ok(socket)
+                                .map(move |plain| Connection::plain(plain, why_no_tls));
```

Comment:
NIT: I think this can be just
```rust
let f = future::ok(Connection::plain(socket, why_no_tls));
```

Reply:
Ah, that makes sense. To me, reading this just makes me think "Ah, this case used to be more complex and some intermediate combinators were removed" or something, but I understand the rationale.